src/lib/ams.c

```diff
@@ -0,0 +1,736 @@
+								   producer_table[i].message_type_id);
+			if (index >= 0) {
+				memcpy_s(&pin_props[count].message_type_uuid, AMS_MESSAGE_UUID_SIZE,
+					 &shared_c->uuid_table[index].message_uuid,
+					 AMS_MESSAGE_UUID_SIZE);
+				count++;
+			} else {
+				tr_err(&ams_tr, "get_output_async_pin_props(): uuid index not found for message_type_id: %u",
+				       producer_table[i].message_type_id);
+			}
+		}
+	}
+
+	ams_release(shared_c);
+	if (err == 0)
+		*output_count = count;
+
+	return err;
+}
+
+static int ams_process_slot(struct async_message_service *ams, uint32_t slot)
+{
+	struct ams_shared_context __sparse_cache *shared_c;
+	struct ams_message_payload msg;
+	uint16_t module_id;
+	uint16_t instance_id;
+
+	shared_c = ams_acquire(ams->ams_context->shared);
+
+	msg = shared_c->slots[slot].u.msg;
+	module_id = shared_c->slots[slot].module_id;
+	instance_id = shared_c->slots[slot].instance_id;
+
+	ams_release(shared_c);
+	tr_info(&ams_tr, "ams_process_slot slot %d msg %d from 0x%08x",
+		slot, msg.message_type_id,
+		msg.producer_module_id << 16 | msg.producer_instance_id);
+
+	return ams_message_send_internal(ams, &msg, module_id, instance_id, slot);
+}
+
+#if CONFIG_SMP
+
+static void ams_task_add_slot_to_process(struct ams_task *ams_task, uint32_t slot)
+{
+	int flags;
+
+	/* only done on main core, irq disabling is enough */
+	irq_local_disable(flags);
+	ams_task->pending_slots |= BIT(slot);
+	irq_local_enable(flags);
+}
+
+int process_incoming_message(uint32_t slot)
+{
+	struct async_message_service *ams = sof_get()->ams;
+	struct ams_task *task = &ams->ams_task;
+
+	ams_task_add_slot_to_process(task, slot);
+
+	return schedule_task(&task->ams_task, 0, 100);
+}
+
+static int process_slot(struct async_message_service *ams, uint32_t slot)
+{
+	return ams_process_slot(ams, slot);
+}
+#endif /* CONFIG_SMP */
+
+/* ams task */
+
+static enum task_state process_message(void *arg)
+{
+	struct ams_task *ams_task = arg;
+	uint32_t slot;
+	int flags;
+
+	if (ams_task->pending_slots == 0) {
+		tr_err(&ams_tr, "Could not process message! Skipping.");
+		return SOF_TASK_STATE_COMPLETED;
+	}
+
+	ams_task->is_in_do_work = true;
+	#ifdef __XCC__
+	slot = 31 - XT_NSAU(ams_task->pending_slots);
+	#else
+	slot = 31 - clz(ams_task->pending_slots);
+	#endif
+
+	process_slot(ams_task->ams, slot);
+
+	/* only done on main core, irq disabling is enough */
+	irq_local_disable(flags);
+	ams_task->pending_slots &= ~BIT(slot);
+	irq_local_enable(flags);
+	ams_task->is_in_do_work = false;
+	schedule_task_cancel(&ams_task->ams_task);
+
+	return SOF_TASK_STATE_COMPLETED;
+}
+
+struct task_ops ams_task_ops = {
+	.run = process_message,
+	.complete = NULL,
+	.get_deadline = NULL,
+};
+
+static int ams_task_init(void)
+{
+	int ret;
+	struct async_message_service *ams = sof_get()->ams;
+	struct ams_task *task = &ams->ams_task;
+
+	task->ams = ams;
+	ret = schedule_task_init_edf(&task->ams_task, SOF_UUID(ams_uuid),
+				     &ams_task_ops, task, 0, 0);
```

Comment:
the task can only run on core 0? How do you process messages on other cores then?

Reply:
Btw, does this need to be a Zephyr thread (with a small stack)?